### ISSUE TITLE
Leave xdebug enabled by default in ddev-webserver (How to enable xdebug for PHPStorm Unit tests using ddev-webserver)

### ISSUE THREAD

**Issue by BobMali** [drud/ddev#2245]
**Describe the bug**
Before ddev 1.14 phpstorm's setting "CLI Interpreter" would show the debugger as installed and the current version (see interpreter1-13 screenshot).
After upgrading to ddev 1.14 the phpstorm setting will always show the debugger as not installed (see interpreter1-14 screenshot) wether xdebug is activated via ddev or not.
Listening for incoming connections and debugging those connections still works, however debugging tests for example does not.

**To Reproduce**
1. Create test-project with ddev 1.13 and start
2. Open test-project in phpstorm and navigate to Settings -> Languages & Frameworks -> PHP -> CLI Interpreter
3. Add CLI Interpreter from docker (use ddev web container for test-project)
4. See debugger as installed
5. ddev poweroff
6. Upgrade ddev to v1.14 and ddev start test-project
7. Repeat step 2 and 3
8. See debugger as not installed

**Expected behavior**
The debugger should be shown as installed so that tests might be debugged again.

**Screenshots**
interpreter1-13
![interpreter1-13](https://user-images.githubusercontent.com/13960257/81390413-b086f300-911b-11ea-8580-1ace1d516124.png)
interpreter1-14
![interpreter1-14](https://user-images.githubusercontent.com/13960257/81390422-b7ae0100-911b-11ea-930b-137201ede50b.png)

**Version and configuration information (please complete the following information):**
 - Host computer OS and Version: Ubuntu 18.04.4
 - Docker version:
```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b7f0
 Built:             Wed Mar 11 01:25:46 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 01:24:19 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```
 - ddev version: 1.13.2 and 1.14.1
 - phpstorm version: 2020.1.1

**Comment by rfay** [comment]
In v1.14 the web image does not have xdebug enabled by default because it slowed the build down. If you wanted it to be enabled using this technique (which I don't recommend) then you can start the web container with the environment variable DDEV_XDEBUG_ENABLED set to true.

But there's no need to even use the php inside the container, and with the technique you're using, PHPStorm actually starts a new, completely separate web container (with defaults, which is why you don't see xdebug)

By doing this:

> Add CLI Interpreter from docker (use ddev web container for test-project)

As far as I know, there's no need to specify a PHP interpreter in PHPStorm. Here's what my configuration looks like:

<img width="758" alt="Preferences" src="https://user-images.githubusercontent.com/112444/81415085-2cc20c00-9105-11ea-980f-f7f47be2071f.png">

**Comment by rfay** [comment]
I took a quick look and was unable to find a way to configure PHPStorm's launching of its own container with an environment variable. PHPStorms's in-container support has always been pretty useless with ddev because PHPStorm launches its own container without any of the things you need, instead of using the one that's already running.

**Comment by BobMali** [comment]
Thanks for your quick answer first at all.

> As far as I know, there's no need to specify a PHP interpreter in PHPStorm. Here's what my configuration looks like:

You are quite right. no interpreter is needed if one is listening for incoming connections.
However, if one is to start a unit test an interpreter is needed. If that unit test is started in debug mode there will be an error because the debugger extension is not recognised/installed. 
Sorry if i did not make myself clear.

Just to make sure i understand you correctly: If i want the previous behaviour i have to start the web container with DDEV_XDEBUG_ENABLED or open an issue with Jetbrains about their in-container support?

**Comment by rfay** [comment]
I explored the unit test capability at some length on https://github.com/drud/ddev/issues/1579#issuecomment-491642099 and never thought anybody would do that again because of the fundamental problem that Jetbrains assumes it is in charge of running its own container.

I think you could probably solve this easily by using a .ddev/web-build/Dockerfile with `ENV  DDEV_XDEBUG_ENABLED=true` in it, then make sure that you tell PHPStorm to use the image named `drud/ddev-webserver:v1.14.1-<projectname>-built`.  You could also just build your own derivative image with the env set.

I don't think it helps much to open an issue with Jetbrains, as what they're doing has nothing to do with what you're trying to do. 

But glad you're pioneering this, and I hope you'll write up how you use it when you get it going. Happy to help you figure it out.

**Comment by rfay** [comment]
But there's also the question... Why would you want xdebug enabled to run unit tests? It will just slow everything down terribly.

**Comment by BobMali** [comment]
You are right. But there are times when debugging is preferred over performance in unit tests.
It is for example easier to follow the data flow through complicated methods of legacy projects if it is done via debugger. At least for me.
Another reason might be failing tests without apparent reason or concrete failure message. This happened to me a few times when using prophecy mocks. Using the debugger revealed slightly different parameters than expected.
Granted those are edge cases but debugging made resolving those cases easier/faster for me.

**Comment by rfay** [comment]
I guess xdebug is super useful when *debugging* tests, absolutely. But I wouldn't expect to be using PHPStorm's test runner for that.

**Comment by rfay** [comment]
BTW, I tried my suggestion of `ENV DDEV_XDEBUG_ENABLED=true` and it didn't work because the start script turns off xdebug. 

But *debugging* with PHPStorm is a completely different thing than running unit tests, I still don't understand why Xdebug helps you with the PHP interpreter to run the tests.

Anyway, the bottom line is, you might be better to use any old PHP image for this, instead of even using the ddev PHP image, because what you're ending up doing has nothing to do with ddev. It's PHPStorm running a container for you. But it will be faster with xdebug turned off anyway.

**Comment by rfay** [comment]
Here's what you need for .ddev/web-build/Dockerfile: 
```
ARG BASE_IMAGE
FROM $BASE_IMAGE
RUN enable_xdebug
```

Then you'll `ddev start` to build the "built" version, in my case that's `drud/ddev-webserver:v1.14.2-d8composer-built`, you can find it by `docker ps | grep ddev-webserver`

Closing this for now, happy to continue the conversation here or elsewhere.

**Comment by NicoHaase** [comment]
I've selected the ddev containers to run unit tests, and using some older versions of ddev, this worked perfect :) If you see any chance of getting this back, I'd love to help find a way. But I haven't dived into the inner stuff of ddev yet, so I don't think I could get this done on my own

**Comment by rfay** [comment]
This is easy enough to remedy, and it seems one or two people use PHPStorm this way, even though it has nothing really to do with ddev. I'll put it on the board for v1.16.

**Comment by rfay** [comment]
Fixed in #2460, will be in ddev v1.16, you'll be able to use this from v1.16-alpha2 on in edge releases.

**Comment by rfay** [comment]
@BobMali @NicoHaase it doesn't seem like this should be necessary any more, the PhpStorm experience has changed so much. As a result this is lined up for reversion in 
* #4599 

Xdebug would no longer be on by default in the ddev-webserver container.

I'd love you to comment there if you have concerns.

**Comment by JBlocks** [comment]
> @BobMali @NicoHaase it doesn't seem like this should be necessary any more, the PhpStorm experience has changed so much. As a result this is lined up for reversion in
> 
> * [Don't enable xdebug by default in ddev-webserver, fixes #4597 #4599](https://github.com/ddev/ddev/pull/4599)
> 
> Xdebug would no longer be on by default in the ddev-webserver container.
> 
> I'd love you to comment there if you have concerns.

After this change, the `enable_xdebug` / `ddev xdebug on` or yaml config `xdebug_enabled` doesn't take affect.
In the config.yaml is `php_version: "8.1"` defined. PhpStorm could not determin the debugger in the Cli interpreters anymore. 
For this reason, a PHPUnit test can no longer be started in debug mode.
The classic HTTP web requests (php-fpm) are easily recognized after an "enable_xdebug".

![image](https://user-images.githubusercontent.com/9034176/229765575-99bc184b-d0c5-4e50-b072-b6f17ec64daa.png)

The only workaround for me is the mentioned [custom Dockerfile](https://github.com/ddev/ddev/issues/2245#issuecomment-628333739) with the `RUN enable_xdebug` line. 
My ddev version is: `v1.21.6`

**Comment by rfay** [comment]
@JBlocks I don't actually know what problem you're having, but it seems like it probably needs a new issue. You should either be using the PhpStorm extension or manually configuring PhpStorm as described in https://ddev.readthedocs.io/en/latest/users/install/phpstorm/
